Extension/src/LanguageServer/devcmd.ts

```diff
@@ -0,0 +1,214 @@
+import * as vscode from 'vscode';
+import * as nls from 'vscode-nls';
+import { isWindows } from '../constants';
+import { CppSettings } from './settings';
+
+nls.config({ messageFormat: nls.MessageFormat.bundle, bundleFormat: nls.BundleFormat.standalone })();
+const localize: nls.LocalizeFunc = nls.loadMessageBundle();
+
+const errorNoContext = localize('no.context.provided', 'No context provided');
+const errorNotWindows = localize('not.windows', 'The "Set Visual Studio Developer Environment" command is only available on Windows');
+const errorNoVSFound = localize('error.no.vs', 'A Visual Studio installation with the C++ compiler was not found');
+export const errorOperationCancelled = localize('operation.cancelled', 'The operation was cancelled');
+const errorNoHostsFound = localize('no.hosts', 'No hosts found');
+const configuringDevEnv = localize('config.dev.env', 'Configuring developer environment...');
+const selectVSInstallation = localize('select.vs.install', 'Select a Visual Studio installation');
+const advancedOptions = localize('advanced.options', 'Advanced options...');
+const advancedOptionsDescription = localize('advanced.options.desc', 'Select a specific host/target architecture, toolset version, etc.');
+const selectToolsetVersion = localize('select.toolset', 'Select a toolset version');
+const selectHostTargetArch = localize('select.host.target', 'Select a host and target architecture');
+
+export function isEnvironmentOverrideApplied(context?: vscode.ExtensionContext) {
+    return context?.environmentVariableCollection.get('VCToolsInstallDir') !== undefined;
+}
+
+export async function setEnvironment(context?: vscode.ExtensionContext) {
+    if (!isWindows) {
+        throw new Error(errorNotWindows);
+    }
+
+    if (!context) {
+        throw new Error(errorNoContext);
+    }
+
+    const vses = await getVSInstallations();
+    if (!vses) {
+        throw new Error(errorNoVSFound);
+    }
+
+    let vs = await chooseVSInstallation(vses);
+    let options: vcvars.Options | undefined;
+    if (!vs) {
+        const compiler = await getAdvancedConfiguration(vses);
+        vs = compiler.vs;
+        options = compiler.options;
+    }
+    const vars = await vscode.window.withProgress({
+        cancellable: false,
+        location: vscode.ProgressLocation.Notification,
+        title: configuringDevEnv
+    }, () => vcvars.getVCVars(vs, options));
+
+    if (!vars || !vars['INCLUDE']) {
+        throw new Error(localize('something.wrong', 'Something went wrong: {0}', JSON.stringify(vars)));
+    }
+
+    const host = vars['VSCMD_ARG_HOST_ARCH'];
+    const target = vars['VSCMD_ARG_TGT_ARCH'];
+    const arch = vcvars.getArchitecture({
+        host: match(host, { 'x86': 'x86', 'x64': 'x64' }) ?? 'x64',
+        target: match(target, { 'x86': 'x86', 'x64': 'x64', 'arm64': 'ARM64', 'arm': 'ARM' }) ?? 'x64'
+    });
+    const settings = new CppSettings(vscode.workspace.workspaceFolders?.at(0)?.uri);
```

Comment:
No workspace uri should be passed to CppSettings here -- it's causing this warning in the extension host logging: `

[warning] [ms-vscode.cpptools] Accessing a window scoped configuration for a resource is not expected. To associate 'C_Cpp.persistVsDeveloperEnvironment' to a resource, define its scope to 'resource' in configuration contributions in 'package.json'

Reply:
fixed